celery_exporter/celery_state.py

```diff
@@ -0,0 +1,169 @@
+from celery import states
+
+CELERY_MISSING_DATA = "undefined"
+# name, state, runtime, queue
+CollectOutcome = Tuple[Optional[str], Optional[str], Optional[float], Optional[str]]
+# name, queue, latency
+LatencyOutcome = Tuple[Optional[str], Optional[str], Optional[float]]
+
+
+def is_task_event(kind: str) -> bool:
+    return "task" in kind
+
+
+class TaskState(Enum):
+    PENDING = states.PENDING
+    RECEIVED = states.RECEIVED
+    STARTED = states.STARTED
+    FAILURE = states.FAILURE
+    RETRY = states.RETRY
+    SUCCESS = states.SUCCESS
+    REVOKED = states.REVOKED
+    REJECTED = states.REJECTED
+    UNDEFINED = "UNDEFINED"
+
+    @classmethod
+    def from_event(cls, event_kind: str) -> "TaskState":
+        MAPPING = {
+            "sent": TaskState.PENDING,
+            "received": TaskState.RECEIVED,
+            "started": TaskState.STARTED,
+            "failed": TaskState.FAILURE,
+            "retried": TaskState.RETRY,
+            "succeeded": TaskState.SUCCESS,
+            "revoked": TaskState.REVOKED,
+            "rejected": TaskState.REJECTED,
+        }
+
+        return MAPPING.get(event_kind, TaskState.UNDEFINED)
+
+
+@dataclass
+class Task:
+    uuid: str = CELERY_MISSING_DATA
+    name: str = CELERY_MISSING_DATA
+    local_received: float = 0.0
+    runtime: Optional[float] = None
+    state: TaskState = TaskState.UNDEFINED
+
+    def update_from_event(self, event: Dict):
+        kind: Optional[str] = event.get("type")
+        if kind is None:
+            raise ValueError("Invalid Event: missing type")
+        elif not (isinstance(kind, str)):
+            raise TypeError(f"Except type str for kind, found {type(kind)}")
+
+        splitted = kind.split("-")
+        state = splitted[1]
+
+        if is_task_event(kind):
+            self.uuid = event.get("uuid", CELERY_MISSING_DATA)
+            self.state = TaskState.from_event(state)
+            local_received = event.get("local_received")
+            if self.local_received is None:
+                raise ValueError("Invalid Event: missing local_received")
+            elif not (isinstance(local_received, float)):
+                raise TypeError(
+                    f"Expected type float for local_received, found {type(local_received)}"
+                )
+            self.local_received = local_received
+
+            self.name = event.get("name", CELERY_MISSING_DATA)
+            self.runtime = event.get("runtime")
+
+
+class CeleryState:
+    def __init__(
+        self,
+        event_count: int,
+        task_count: int,
+        queue_by_task: Dict[str, str],
+        tasks: LRUCache[str, Task],
+    ):
+        self.event_count = event_count
+        self.task_count = task_count
+        self.queue_by_task = queue_by_task
+        self.tasks = tasks
+
+    @classmethod
+    def new(cls, max_tasks_in_memory: int) -> "CeleryState":
+        return cls(
+            event_count=0,
+            task_count=0,
+            queue_by_task=defaultdict(),
```

Comment:
Does this need to be a defaultdict? I don't _think_ so but might be missing something. We should change the type annotation if so.

Reply:
You're right, changing to `dict`